Grid based density and source position sampling

JulienDoerner opened the pull request:
This PR has 3 major parts.

1. An interface to use a Grid1f  as a mass distribution is introduced.
2. A grid feature to restrict a grid to the volume defined by the origin, points and spacing is introduced. This is necessary for mass distribution grids, which should not be repeated.  
3. A source feature for sampling the source position from a given mass distribution via Monte Carlo sampling.

Additionally, all aspects are shown in a new example notebook which uses data for the mass distribution from Mertsch+(2021). This data are added to the additional resources (https://desycloud.desy.de/index.php/s/6zpYsCd29RLbFiA).

lukasmerten commented on include/crpropa/Grid.h:
```diff
@@ -215,6 +219,10 @@ class Grid: public Referenced {
 		reflective = b;
 	}
 
+	void setClipVolume(bool b) {
```
Add some docstring

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/Source.h:
```diff
@@ -816,6 +817,73 @@ class SourceTag: public SourceFeature {
+	@class SourceMassDistribution
+	@brief	Source position follows a given mass distribution
+
+	The (source)position of the candidate is sampled from a give mass distribution. The distribution uses the getDensity function of the density module. 
```
[...] from a given [...]

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/Source.h:
```diff
@@ -816,6 +817,73 @@ class SourceTag: public SourceFeature {
+	@brief	Source position follows a given mass distribution
+
+	The (source)position of the candidate is sampled from a give mass distribution. The distribution uses the getDensity function of the density module. 
+	If a weighting for different components is desired, the use of different densities in a densityList is recomended.
```
recommended

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/Source.h:
```diff
@@ -816,6 +817,73 @@ class SourceTag: public SourceFeature {
+
+public: 
+	/** Constructor
+	@param density: CRPropa mass distrubiton 
```
distribution

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/Source.h:
```diff
@@ -816,6 +817,73 @@ class SourceTag: public SourceFeature {
+	/** Constructor
+	@param density: CRPropa mass distrubiton 
+	@param maxDensity:	maximal density in the region where the position should be sampled
+	@param x:	the position will be sampled in the range [-x, x]. Non symetric values can be set with setXrange.
```
[...] symmetric [...]

also at occurrences

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/Source.h:
```diff
@@ -816,6 +817,73 @@ class SourceTag: public SourceFeature {
+
+	void prepareParticle(ParticleState &particle) const;
+
+	/** Set the maximal density in the region of interesst. This parameter is necesarry for the sampling
```
interest [...] necessary

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/Source.h:
```diff
@@ -816,6 +817,73 @@ class SourceTag: public SourceFeature {
+	*/
+	void setXrange(double xMin, double xMax);
+
+	/** set yx-range in which the position of the candidate will be sampled. y in [yMin, yMax].
```
y-range

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/massDistribution/Massdistribution.h:
```diff
@@ -38,7 +39,8 @@ class DensityList: public Density {
 	/** Get HII density at a given position.
 	 @param position position in galactic coordinates with Earth at (-8.5 kpc, 0, 0)
-	 @returns Density of HII at given position in parts/m^3, sum up all HII densities from added densities */
+	 @returns Density of HII at given position in parts/m^3, sum up all HII densities from added densities 
```
let's be more precise 
particles / m^3

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/massDistribution/Massdistribution.h:
```diff
@@ -51,10 +53,84 @@ class DensityList: public Density {
+ @class DensityGrid
+ @brief Wrapper to use a Grid1f for a density
+
+ The DensityGrid uses a given grid for the choosen density type. More than one type can be choosen to follow the same distribution.
```
chosen

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/massDistribution/Massdistribution.h:
```diff
@@ -51,10 +53,84 @@ class DensityList: public Density {
+ @brief Wrapper to use a Grid1f for a density
+
+ The DensityGrid uses a given grid for the choosen density type. More than one type can be choosen to follow the same distribution.
+ If no type is choosen a warning will be raised and all densities are 0.
```
chosen

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/massDistribution/Massdistribution.h:
```diff
@@ -51,10 +53,84 @@ class DensityList: public Density {
+	double getHIIDensity(const Vector3d &position) const;
+	
+	/** Get H2 density at a given position.
+	 @param position position in galactic coordinates with Earth at (-8.5 kpc, 0, 0)
```
Galactic coordinates

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/massDistribution/Massdistribution.h:
```diff
@@ -51,10 +53,84 @@ class DensityList: public Density {
 
-#endif  // CRPROPA_MASSDISTRIBUTION_H
+	/** Get density at a given position.
+	 @param position position in galactic coordinates with Earth at (-8.5 kpc, 0, 0)
```
Galactic

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/massDistribution/Massdistribution.h:
```diff
@@ -51,10 +53,84 @@ class DensityList: public Density {
-#endif  // CRPROPA_MASSDISTRIBUTION_H
+	/** Get density at a given position.
+	 @param position position in galactic coordinates with Earth at (-8.5 kpc, 0, 0)
+	 @returns Density in parts/m^3, sum up densities from added densities 
```
particles / m^3

JulienDoerner replied:
done

lukasmerten commented on include/crpropa/massDistribution/Massdistribution.h:
```diff
@@ -51,10 +53,84 @@ class DensityList: public Density {
+	/** Get the density of nucleons.
+	 This is the number of nucleons per volume, summed up all activated density and weight molecular hydrogyen twice
+	 @param position position in galactic coordinates with Earth at (-8.5 kpc, 0, 0)
+	 @returns Density of nucleons at given position in parts/m^3, sum up all nucleon densities from added densities 
```
particles / m^3

JulienDoerner replied:
done

lukasmerten commented:
I have not yet the code itself, but focused on the docstrings.

reichherzerp commented on include/crpropa/Grid.h:
```diff
@@ -262,6 +271,16 @@ class Grid: public Referenced {
 	  routine with the setInterpolationType function.*/
 	T interpolate(const Vector3d &position) {
+	// check for volume
+	if(clipVolume) {
```
Space between statement and parenthesis

reichherzerp commented on include/crpropa/Grid.h:
```diff
@@ -262,6 +271,16 @@ class Grid: public Referenced {
+		bool isInVolume = (position.x >= origin.x) && (position.x <= edge.x);
+		isInVolume &= (position.y >= origin.y) && (position.y <= edge.y);
+		isInVolume &= (position.z >= origin.z) && (position.z <= edge.z);
+		if(!isInVolume) 
```
Space between statement and parenthesis: if (....

reichherzerp commented on src/Source.cpp:
```diff
@@ -1116,4 +1116,87 @@ void SourceTag::setTag(std::string tag) {
+	density(density), maxDensity(max), xMin(-x), xMax(x), yMin(-y), yMax(y), zMin(-z), zMax(z) {}
+
+void SourceMassDistribution::setMaximalDensity(double maxDensity) {
+	if(maxDensity <= 0) {
```
if (

reichherzerp commented on src/Source.cpp:
```diff
@@ -1116,4 +1116,87 @@ void SourceTag::setTag(std::string tag) {
+}
+
+void SourceMassDistribution::setXrange(double xMin, double xMax) {
+	if(xMin > xMax){
```
if (....

reichherzerp commented on src/Source.cpp:
```diff
@@ -1116,4 +1116,87 @@ void SourceTag::setTag(std::string tag) {
+}
+
+void SourceMassDistribution::setZrange(double zMin, double zMax) {
+	if(zMin > zMax){
```
if (....

reichherzerp commented on src/Source.cpp:
```diff
@@ -1116,4 +1116,87 @@ void SourceTag::setTag(std::string tag) {
+	Vector3d pos; 
+	Random &rand = Random::instance();
+
+	for(int i = 0; i < maxTries; i++) {
```
for (....

reichherzerp commented on src/massDistribution/Massdistribution.cpp:
```diff
@@ -41,5 +41,106 @@ double DensityList::getNucleonDensity(const Vector3d &position) const {
+
+void DensityGrid::checkAndWarn() {
+	bool allDeactivated = (isForHI == false) & (isForHII == false) & (isForH2 == false);
+	if(allDeactivated) {
```
There are several positions where it should be `if (....`

reichherzerp commented on src/Source.cpp:
```diff
@@ -1116,4 +1116,87 @@ void SourceTag::setTag(std::string tag) {
+
+void SourceMassDistribution::setXrange(double xMin, double xMax) {
+	if(xMin > xMax){
+		KISS_LOG_WARNING << "SourceMassDistribution: minimal x-value must be smaler than the maximal one\n";
```
... does not exceed the maximal one

reichherzerp replied:
(also for the other components)

JulienDoerner replied:
I changed to "must not exceed ... "

reichherzerp commented on include/crpropa/Grid.h:
```diff
@@ -262,6 +271,16 @@ class Grid: public Referenced {
+	if(clipVolume) {
+		Vector3d edge = origin + Vector3d(Nx, Ny, Nz) * spacing;
+		bool isInVolume = (position.x >= origin.x) && (position.x <= edge.x);
+		isInVolume &= (position.y >= origin.y) && (position.y <= edge.y);
```
For consistency, you can consider using also `&&=` instead of `&=`

reichherzerp commented on src/massDistribution/Massdistribution.cpp:
```diff
@@ -41,5 +41,106 @@ double DensityList::getNucleonDensity(const Vector3d &position) const {
+	}
+
+void DensityGrid::checkAndWarn() {
+	bool allDeactivated = (isForHI == false) & (isForHII == false) & (isForH2 == false);
```
For **consistency**, you may want to use also `&&` as you do at another part of the code (or vice versa)

reichherzerp commented on include/crpropa/Grid.h:
```diff
@@ -262,6 +271,16 @@ class Grid: public Referenced {
 	  By default this it the trilinear interpolation. The user can change the
 	  routine with the setInterpolationType function.*/
 	T interpolate(const Vector3d &position) {
+	// check for volume
```
the whole section between `T interpolate(const Vector3d &position) {` needs an indent.

reichherzerp commented on include/crpropa/Grid.h:
```diff
@@ -262,6 +271,16 @@ class Grid: public Referenced {
+		isInVolume &= (position.y >= origin.y) && (position.y <= edge.y);
+		isInVolume &= (position.z >= origin.z) && (position.z <= edge.z);
+		if(!isInVolume) 
+			return T(0.);
```
Would a KISS message help the user to see that the current position is outside the grid?

JulienDoerner replied:
I would not do it. In a full simulation where it is possible that often the interpolation is called outside a grid that causes a lot of writing to the console which makes the simulation extremely slow and other important warnings can easily be overseen.

reichherzerp reviewed (changes requested):
Overall it looks good to me. I requested some changes to improve consistency and to follow the style guidelines.

reichherzerp commented on src/Source.cpp:
```diff
@@ -1130,26 +1130,26 @@ void SourceMassDistribution::setMaximalDensity(double maxDensity) {
 void SourceMassDistribution::setXrange(double xMin, double xMax) {
-	if(xMin > xMax){
-		KISS_LOG_WARNING << "SourceMassDistribution: minimal x-value must be smaler than the maximal one\n";
+	if(xMin > xMax) {
```
if (

reichherzerp commented on src/Source.cpp:
```diff
@@ -1130,26 +1130,26 @@ void SourceMassDistribution::setMaximalDensity(double maxDensity) {
 void SourceMassDistribution::setYrange(double yMin, double yMax) {
-	if(yMin > yMax){
-		KISS_LOG_WARNING << "SourceMassDistribution: minimal y-value must be smaler than the maximal one\n";
+	if(yMin > yMax) {
```
if (

reichherzerp commented on src/Source.cpp:
```diff
@@ -1130,26 +1130,26 @@ void SourceMassDistribution::setMaximalDensity(double maxDensity) {
 void SourceMassDistribution::setZrange(double zMin, double zMax) {
-	if(zMin > zMax){
-		KISS_LOG_WARNING << "SourceMassDistribution: minimal z-value must be smaler than the maximal one\n";
+	if(zMin > zMax) {
```
if (

JulienDoerner commented:
I added the code style comments by @reichherzerp.

JulienDoerner commented:
I also added some test for the new feature and added the changes to the CHANGELOG.md

reichherzerp reviewed (changes requested):
Thanks for implementing the changes; it looks good now, apart from the error in the test Grid3f.Interpolation with none `SIMD_EXTENSIONS`. The error message in the `LastTest.log`
file looks like:

[----------] 6 tests from Grid3f
[ RUN      ] Grid3f.Interpolation
unknown file: Failure
C++ exception with description "Tried to use tricubic Interpolation without SIMD_EXTENSION. SIMD Optimization is neccesary for tricubic interpolation of vector grids.
" thrown in the test body.
[  FAILED  ] Grid3f.Interpolation (0 ms)
[ RUN      ] Grid3f.Periodicity
[       OK ] Grid3f.Periodicity (0 ms)
[ RUN      ] Grid3f.Reflectivity
[       OK ] Grid3f.Reflectivity (0 ms)
[ RUN      ] Grid3f.DumpLoad
[       OK ] Grid3f.DumpLoad (0 ms)
[ RUN      ] Grid3f.DumpLoadTxt
[       OK ] Grid3f.DumpLoadTxt (0 ms)
[ RUN      ] Grid3f.Speed
[       OK ] Grid3f.Speed (2 ms)
[----------] 6 tests from Grid3f (2 ms total)

reichherzerp reviewed (approved):
The test error is only appearing on Macbooks and seems to be related to the error in testInteraction. As there are attempts to work on this issue in another PR, I'll approve this PR.

JulienDoerner commented:
I also changed the warning behavior of the densities. Until now, warnings are given if one density type is not activated. This causes a lot of warnings if for example only one component of the gas is used. 
I changed the warnings to occur only if all types are deactivated and the density is accordingly 0.

reichherzerp commented:
That's a good idea, and I'll proceed to merge now. Thanks for the contribution!
